### PR TITLE
Add passing and breaking spec for nested model assignment

### PR DESCRIPTION
**Pull request by machty**
...via json_attributes

Connects #122

The non-polymorphic approach seems to work, but
not the polymorphic version.

### CODE REVIEW

**Comment by machty** [comment]
Noticed that my breaking polymorphic test was omitting the "type" fields; just pushed another commit to add it in. The reason I omitted at first is I was generated my test data via `.json_attributes.to_json`, which seems to omit the `"type": "Model2"` fields, whereas those type fields are present on `.json_attributes_before_type_cast`

**Comment by jrochkind** [comment]
Thanks @machty this is super helpful, trying to take a look today.

**Comment by jrochkind** [comment]
And yeah... I see now that #to_json won't work, as you discovered. That really annoys me. i don't like making you do that `json_attributes_before_type_cast` nonsense... but I see how it is in some ways inherent in the current architecture. 

This gives me more ideas for some potential changes in an attr_json 2.x though. But for the meantime, we should hopefully at least be able to get your submitted failing test passing.

**Comment by jrochkind** [comment]
Hey @machty , can you give me permissions to write to your branch, so I can make some changes here in this PR?  Then we can keep things neat and actually finish here on this PR, instead of me having to close this and open up a new one to make the fix and CHANGELOG entry. 

https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by machty** [comment]
I was hasty and just added you as a collaborator, either way i think you should have access now.

**Comment by jrochkind** [comment]
I can't get github to let me push to your branch. 🤷   I guess I'll just close this one and open up my own branch?

**Comment by jrochkind** [comment]
oh wait i think i know let me try again

**Comment by jrochkind** [comment]
OK, that was a simple fix!  (you did the hard part of reproducing in a test!). I think it makes sense. 

I'm going to try to spend today on attr_json, and release a new version including this fix. Thanks!

**Comment by machty** [comment]
Thanks for working on this; I was successfully able to use json_attr master to implement my ActiveAdmin form, using `json_attributes_before_type_cast` so that the `type` is included.

**Comment by jrochkind** [comment]
@machty Awesome! Thanks for bringing it to my attention.

Another alternative to `json_attributes_before_type_cast` that *might* make more sense in this case, might be actually using the serializer... something like, hmm, how about this....

    model.class.attribute_types["json_attributes"].serialize(model.json_attributes)

Hmm.. I thought that would get me an actual Hash instead of a serialized hash, but that gives me a serialized hash (String) too! I think I still like it better than `json_attributes_before_type_cast` -- at least it will be based on what's currently in memory, rather than "the last thing we fetched from the db". 

I dunno, I don't love the way this works for your use case, but it's good you can at least made it work!
